test/src/test-cppapi-ndrectangle.cc

```diff
@@ -77,3 +71,26 @@ TEST_CASE_METHOD(
   CHECK(range[0] == 30);
   CHECK(range[1] == 40);
 }
+
+TEST_CASE("NDRectangle - Errors", "[cppapi][ArraySchema][NDRectangle]") {
+  tiledb::Context ctx;
+
+  // Create a domain
+  tiledb::Domain domain(ctx);
+  auto d1 = tiledb::Dimension::create<int32_t>(ctx, "d1", {{1, 10}}, 5);
+  auto d2 = tiledb::Dimension::create(
+      ctx, "d2", TILEDB_STRING_ASCII, nullptr, nullptr);
+  domain.add_dimension(d1);
+  domain.add_dimension(d2);
+
+  // Create an NDRectangle
+  tiledb::NDRectangle ndrect(ctx, domain);
+
+  // Set range with non-existent dimension
+  CHECK_THROWS(ndrect.set_range(2, 1, 2));
+  CHECK_THROWS(ndrect.set_range("d3", 1, 2));
+
+  // Set range out of order
+  CHECK_THROWS(ndrect.set_range(0, 2, 1));
+  CHECK_THROWS(ndrect.set_range("d2", "bbb", "aaa"));
```

Comment:
I tried adding a check that setting a string to an integer dimension and vice versa fails, but such validation is not possible in the Core side because it does not receive typing information from the user (I added another check locally that in fixed-sized dimensions the size of the range == `sizeof(T) * 2`, but it fails if I pass two four-character strings).

We could add such type validation on the C++ API side, but that would require keeping around the `Domain` passed to the constructor. 🤔